plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java

```diff
@@ -0,0 +1,321 @@
+            return Optional.of(doubleColumnMapping());
+        }
+        if (typeHandle.getJdbcType() == Types.DECIMAL) {
+            int precision = min(typeHandle.getColumnSize().orElse(0), Decimals.MAX_PRECISION);
+            int decimalDigits = typeHandle.getDecimalDigits()
+                    .orElseThrow(() -> new IllegalStateException("decimal digits not present"));
+            int scale = min(decimalDigits, getDecimalDefaultScale(session));
+            return Optional.of(decimalColumnMapping(createDecimalType(precision, scale),
+                    getDecimalRoundingMode(session)));
+        }
+        if (jdbcTypeName.equals("String")) {
+            return Optional.of(varcharColumnMapping(createUnboundedVarcharType()));
+        }
+        if (jdbcTypeName.equals("Date")) {
+            return Optional.of(dateColumnMapping());
+        }
+        if (jdbcTypeName.equals("DateTime")) {
+            // clickhouse not implemented for type=class java.time.LocalDateTime
+            // TODO replace it using timestamp relative function after clickhouse adds support for LocalDateTime
+            return Optional.of(varcharColumnMapping(createUnboundedVarcharType()));
+        }
+        if (getUnsupportedTypeHandling(session) == CONVERT_TO_VARCHAR) {
+            return mapToUnboundedVarchar(typeHandle);
+        }
+        // unknown type ?
+        return Optional.empty();
+    }
+
+    @Override
+    public WriteMapping toWriteMapping(ConnectorSession session, Type type)
+    {
+        if (type == BOOLEAN) {
+            return WriteMapping.booleanMapping("UInt8", booleanWriteFunction());
+        }
+        if (type == TINYINT) {
+            return WriteMapping.longMapping("Int8", tinyintWriteFunction());
+        }
+        if (type == SMALLINT) {
+            return WriteMapping.longMapping("Int16", smallintWriteFunction());
+        }
+        if (type == INTEGER) {
+            return WriteMapping.longMapping("Int32", integerWriteFunction());
+        }
+        if (type == BIGINT) {
+            return WriteMapping.longMapping("Int64", bigintWriteFunction());
+        }
+        if (type == REAL) {
+            return WriteMapping.longMapping("Float32", realWriteFunction());
+        }
+        if (type == DOUBLE) {
+            return WriteMapping.doubleMapping("Float64", doubleWriteFunction());
+        }
+        if (type instanceof DecimalType) {
+            DecimalType decimalType = (DecimalType) type;
+            String dataType = format("decimal(%s, %s)", decimalType.getPrecision(), decimalType.getScale());
+            if (decimalType.isShort()) {
+                return WriteMapping.longMapping(dataType, shortDecimalWriteFunction(decimalType));
+            }
+            return WriteMapping.sliceMapping(dataType, longDecimalWriteFunction(decimalType));
+        }
+        if (type instanceof CharType || type instanceof VarcharType) {
+            return WriteMapping.sliceMapping("String", varcharWriteFunction());
```

Comment:
char values need to be padded with spaces for this to be correct
see `io.trino.spi.type.Chars#padSpaces(io.airlift.slice.Slice, io.trino.spi.type.CharType)`

there is no true char type?

Reply:
ClickHouse no char type

Reply:
Thanks for confirming. Still would be more correct to pad the values on write.

Reply:
Could `FixedString` be applicable for `CharType`?

Reply:
The length of `FixedString` is calculated by the number of ASCII characters, like the following:

```sql
-> create table tbl_fs (col1 FixedString(4)) ENGINE=TinyLog;

CREATE TABLE tbl_fs
(
    `col1` FixedString(4)
)
ENGINE = TinyLog


-> insert into tbl_fs values('1234');

INSERT INTO tbl_fs VALUES

-> insert into tbl_fs values('你好');

INSERT INTO tbl_fs VALUES


Exception on client:
Code: 131. DB::Exception: String too long for type FixedString(4)

Connecting to database default at 10.60.242.112:19000 as user default.
Connected to ClickHouse server version 20.3.5 revision 54433.

-> select * from tbl_fs;

SELECT *
FROM tbl_fs

┌─col1─┐
│ 1234 │
└──────┘
```